File: problem_solving/greedy/py/1029_two_city_scheduling.py
'''
https://leetcode.com/problems/two-city-scheduling/
'''

def two_city_scheduling(costs: list[list[int]]) -> int:
    costs.sort(key=lambda x:(x[0]-x[1])) 
    total_cost = 0
    N = len(costs) // 2

    for i in range(N):
        total_cost += costs[i][0]

    for i in range(N, 2*N):
        total_cost += costs[i][1]

    return total_cost





costs = [[10,20],[30,200],[400,50],[30,20]]
#Output: 110
result = two_city_scheduling(costs)
print("results = " + str(result))



costs = [[259,770],[448,54],[926,667],[184,139],[840,118],[577,469]]
#Output: 1859
result = two_city_scheduling(costs)
print("results = " + str(result))



costs = [[515,563],[451,713],[537,709],[343,819],[855,779],[457,60],[650,359],[631,42]]
#Output: 3086#
result = two_city_scheduling(costs)
print("results = " + str(result))

File: problem_solving/greedy/cpp/1029_two_city_scheduling.cpp
/*
https://leetcode.com/problems/two-city-scheduling/
*/

#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
#include<vector>
#include<algorithm>
using namespace std;

int two_city_minimum_cost(std::vector<vector<int>>& costs)  {
    vector<std::pair<int, int>> diffs;
    int len = costs.size();
    for(int i = 0; i < len; ++i) {
        diffs.push_back({costs[i][0] - costs[i][1], i} );
    }
    //print_vector_of_pairs(diffs);
    std::sort(diffs.begin(), diffs.end());
    //print_vector_of_pairs(diffs);
    int total_cost = 0;
    int n = costs.size() / 2;
    //to palce A
    for(int i = 0; i < n; ++i) {
        total_cost += costs[diffs[i].second][0];
    }
    //to place B
    for(int i = n; i < 2*n; ++i) {
        total_cost += costs[diffs[i].second][1];
    }
    return total_cost;
}


int main() {
    std::vector<std::vector<int>> costs;
    costs = {{10,20},{30,200},{400,50},{30,20}};
    int result = two_city_minimum_cost(costs);
    std::cout << "result: " << result << std::endl;
}